remotes/docker/fetcher.go

```diff
@@ -210,5 +213,25 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
 		}
 	}
 
-	return resp.Body, nil
+	body := resp.Body
+	encoding := resp.Header.Get("Content-Encoding")
+	switch encoding {
+	case "zstd":
+		r, err := zstd.NewReader(body)
+		if err != nil {
+			return nil, err
+		}
+		body = r.IOReadCloser()
+	case "gzip":
+		body, err = gzip.NewReader(body)
+		if err != nil {
+			return nil, err
+		}
+	case "":
+		// no content-encoding applied, use raw body
+	default:
+		return nil, errors.New("unsupported Content-Encoding: " + encoding)
```

Comment:
Looks like `Content-Encoding` can return multiple values (comma-separated) if multiple are applied; https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding

So we may want to either "log" this was a warning, and "assume" the old behavior, or handle those (not exactly sure how to handle "multiple")

Reply:
multiple is indeed possible according to the HTTP RFC, but I can't imagine a practical usage for it. This could be implemented by applying algorithms in reverse order, but do we need such complexity for a pure theoretical corner case? Then returning error here should be enough, or would you like an additional log for it?

Reply:
Yeah, I don't know how common it is; perhaps others know? Wondering if (e.g.) a CDN could apply compression to already compressed streams, in which case perhaps multiple compressions happen.

Thought I'd at least call it out so that it doesn't go overlooked. We could add a comment describing that that's something that's possible.

Reply:
added support for multiple encoding by [e4cd281](https://github.com/containerd/containerd/pull/7563/commits/e4cd2816fdd2c1765d8f52f46dfa39f6d46d42d4)